modules/gapi/include/opencv2/gapi/infer.hpp

```diff
@@ -395,35 +439,144 @@ struct Generic { };
+                        cv::GKinds& kinds)
+{
+    for (auto&& p : blobs) {
+        names.emplace_back(p.first);
+        switch (p.second.index()) {
+            case cv::GInferInputs::InferInput::index_of<cv::GMat>():
+                args.emplace_back(cv::util::get<cv::GMat>(p.second));
+                kinds.emplace_back(cv::detail::OpaqueKind::CV_MAT);
+                break;
+            case cv::GInferInputs::InferInput::index_of<cv::GFrame>():
+                args.emplace_back(cv::util::get<cv::GFrame>(p.second));
+                kinds.emplace_back(cv::detail::OpaqueKind::CV_UNKNOWN);
+                break;
+            default:
+                GAPI_Assert(false);
+        }
+    }
+}
+
+template<typename T = Generic> GInferOutputs
+infer(const std::string& tag, const GInferInputs& inputs)
+{
+    std::vector<cv::GArg> args;
+    std::vector<std::string> names;
+    cv::GKinds kinds;
+
+    unpackBlobs(inputs.getBlobs(), args, names, kinds);
+
+    auto call = makeCall<GInferBase>(tag,
+                                     std::move(args),
+                                     std::move(names),
+                                     std::move(kinds));
 
     return GInferOutputs{std::move(call)};
 }
 
+template<typename T = Generic> GInferOutputs
+infer(const std::string& tag, const cv::GOpaque<cv::Rect>& roi, const GInferInputs& inputs)
+{
+    std::vector<cv::GArg> args;
+    std::vector<std::string> names;
+    cv::GKinds kinds;
+
+    args.emplace_back(roi);
+    kinds.emplace_back(cv::detail::OpaqueKind::CV_RECT);
+
+    unpackBlobs(inputs.getBlobs(), args, names, kinds);
+
+    auto call = makeCall<GInferROIBase>(tag,
+                                        std::move(args),
+                                        std::move(names),
+                                        std::move(kinds));
+
+    return GInferOutputs{std::move(call)};
+}
+
+template<typename T = Generic> GInferListOutputs
+infer(const std::string& tag, const cv::GArray<cv::Rect>& rois, const GInferInputs& inputs)
+{
+    std::vector<cv::GArg> args;
+    std::vector<std::string> names;
+    cv::GKinds kinds;
+
+    args.emplace_back(rois);
+    kinds.emplace_back(cv::detail::OpaqueKind::CV_RECT);
+
+    unpackBlobs(inputs.getBlobs(), args, names, kinds);
+
+    auto call = makeCall<GInferListBase>(tag,
+                                         std::move(args),
+                                         std::move(names),
+                                         std::move(kinds));
+
+    return GInferListOutputs{std::move(call)};
+}
+
+template<typename T = Generic, typename Input>
+typename std::enable_if<cv::detail::accepted_infer_types<Input>::value, GInferListOutputs>::type
+infer2(const std::string& tag,
+       const Input& in,
+       const cv::GInferListInputs& list)
+{
+    std::vector<cv::GArg> args;
+    std::vector<std::string> names;
+    cv::GKinds kinds;
+
+    args.emplace_back(in);
+    auto k = cv::detail::GOpaqueTraits<Input>::kind;
```

Comment:
@dmatveev @smirnov-alexey 
If I use `kinds.push_back(cv::detail::GOpaqueTraits<Input>::kind)` 
I get `undefined reference`

Reply:
Can you show the exact error? Doesn't seem like a template issue to me (e.g. that you should use `.template push_back()`). Have you tried `emplace_back`?

Reply:
https://stackoverflow.com/questions/48748691/why-cant-push-static-const-member-into-vector

Reply:
Seems like that object/field is not created when you're trying to push it according to the answers. `emplace` is not working either?

Reply:
Yes